uClinux-dist/lib/libpam/doc/man/pam_strerror.3
.\" Hey Emacs! This file is -*- nroff -*- source.
.\" ripped off from Rick Faith's getgroups man page
.\" $Id: pam_strerror.3,v 1.3 2005/08/16 03:41:37 magicyang Exp $
.\" Copyright (c) Andrew G. Morgan 1996-7 <morgan@linux.kernel.org>
.TH PAM_STRERROR 3 "1999 Oct 4" "Linux-PAM 0.70" "Programmers' Manual"
.SH NAME

pam_strerror \- return a textual description of a Linux-PAM error

.SH SYNOPSIS
.B #include <security/pam_appl.h>
.br
or,
.br
.B #include <security/pam_modules.h>
.sp
.BI "const char * pam_strerror( pam_handle_t " "*pamh" ", int " pam_error ");"
.sp 2
.SH DESCRIPTION
.B pam_strerror

This function returns some text describing the
.BR Linux-PAM
error associated with the
.B pam_error
argument.

.SH "RETURN VALUE"

On success this function returns a description of the indicated
error.  Should the function not recognize the error, ``Unknown
Linux-PAM error'' is returned.

.SH "CONFORMING TO"
DCE-RFC 86.0, October 1995.

.SH BUGS
.sp 2
This function should be internationalized.

.SH "SEE ALSO"

.BR pam "(8). "

Also, see the three
.BR Linux-PAM
Guides, for
.BR "System administrators" ", "
.BR "module developers" ", "
and
.BR "application developers" ". "


uClinux-dist/release_notes
Release Notes for uClinux for Blackfin Version 2005R4

Version: 2005R4

Date: Dec. 30, 2005
Development Environments

Tool Chain: Toolchain Release 2005R4

u-boot: uboot_1.1.3 Release 2005R4

Host platform: Suse Linux 9.2 or above

Target board: STAMP Board

Note: However, other similar host platforms are also supported, but they are not well tested by now.
Deliverables

Source files

uClinux_dist_2005R4.tar.bz2

Linux ELF file

linux-bf533
linux-bf537
linux-bf561

Compressed Linux image

uImage-bf533
uImage-bf537
uImage-bf561

This document

release_notes_2005R4.pdf

Compressed archive of test results

test_results_bf533_2005R4.tar.gz
test_results_bf537_2005R4.tar.gz
test_results_bf561_2005R4.tar.gz

Summary of test results

test_results_summary_2005R4

Fixes in release 2005R4

See tracker at:

http://blackfin.uclinux.org/tracker/index.php?group_id=17&atid=141
Improvements

   1.
      Support BF561-EZKIT and add a tool to load code to run in core B.
   2.
      Support Bluetechnix CM Platform, which is also based on BF533.
   3.
      Enable module support in default configuration and generate module dependency file for modprobe.
   4.
      Upgrade to mainline source code
          *
            Upgrade uClibc to version 0.9.27.
          *
            Upgrade Linphone to version 1.1.0
          *
            Upgrade LTP to 20051205 release.
   5.
      Reorganize the scheduling and signal handling.
          *
            Defer scheduling and signal handling from different interrupt handlers to interrupt level 15.
          *
            Deal with the deferred exception handler and the defered interrupt handler in the same level of system call handler.
          *
            Run schedule and do signal from each interrupt handler above level 14.
          *
            Move softirq handler to interrupt level 14
   6.
      Prevent double fault exception.
          *
            Make sure that in an exception handler, we use a stack that is covered by a CPLB.
          *
            For exceptions other than CPLB misses, lower our priority to IRQ5, where we can use the normal kernel stack and use other regular kernel functions.
   7.
      Port new applications, libraries and tools
          *
            samba
          *
            strace
          *
            SQLite
          *
            lmbench3.
          *
            whetstone.
          *
            network audio player
          *
            ncurses
   8.
      Work around new hardware anomalys.
          *
            Bypass the first DCLPB fault exception, only handle the second one if it is from the same fault address.
          *
            Do not put SSYNC,CSYNC or IDLE in the second to last instruction of a hardware loop.
          *
            Do not use DMA_RUN bit to detect completion of a DMA transfer.
   9.
      Change paths from bfinnommu to blackfin, per request from Linux Kernel Mailing list.
  10.
      Enable oprofiler support in kernel and Port oprofiler tools
  11.
      Enable interrupt latency benchmark
  12.
      Add new device drivers
          *
            Enable IDE driver
          *
            Enable 802.11 WLAN driver
          *
            Enable PPI driver
          *
            Enable NAND flash driver
          *
            Port YAFFS driver
          *
            Add a common SPI channel driver. All high level drivers access SPI device by this driver.
  13.
      ADI1836 audio driver improvement
          *
            Reduce dummy buffer size from 256 bytes to 4 bytes
          *
            No interrupts generated in dummy DMA transfer
          *
            When audio driver is not active, DMA is also stopped.
          *
            Enable I2S mode for audio input and output. Modify mixer for I2S mode.
          *
            Enable multichannel output. Add a mixer function to change the output channels when play stereo.
          *
            Enable play multi independent audio streams concurrently.
          *
            Make audio DMA stable.
  14.
      Kernel debugging improvement
          *
            Enable hardware single stepping for application debugging.
          *
            Report precise hardware error information
          *
            debug kernel start. Install a mini-execption handler (replacing the U-Boot one), turn the UART on, and dump out the SEQSTAT register.
          *
            Enable hardware trace buffer in dump
  15.
      Improve power management
          *
            Implement the suspend function in bf533_pm_suspend.
          *
            SDRAM enters self-refresh mode before deepsleep to save more power.
          *
            Enable PM on BF537
  16.
      Enable PPPoE
  17.
      Enable DMA in SPI driver.
  18.
      Add board specific Kconfig Option for PF port/pin, which is used for flash and ethernet sharing.
  19.
      Enhance msh to support funtions in scripts.
  20.
      Transmite bfin atomic testandset operation in exception 4 into a new syscall, sys_bfin_spinlock.

Known Issues

A full list of known issues can be found at:

http://blackfin.uclinux.org/tracker/index.php?group_id=17&atid=141
No 	Issue Title
891	tcp stress testing echo service
900	busybox/msh,cd,tar and pipes
960	NFS unstable under heavy load
991	Infrequent audio crash
1012	After running dpmc to_deepsleep testcase, the kernel would panic after a while
1014	kernel built with audio configuration sometimes would fail to boot up and hangs

There are also some issues in the LTP test cases. They are recorded as bug 532, 561, 567, 732, 735, 742, 744, 745, 890 and 1010.
Build Procedure

1. Install Toolchain Release 2005R4

Go to http://blackfin.uclinux.org/frs/?group_id=18 for more information


uClinux-dist/lib/libpam/libpam/pam_strerror.c
/* pam_strerror.c */

/*
 * $Id: pam_strerror.c,v 1.3 2005/08/16 03:41:37 magicyang Exp $
 */

#include "pam_private.h"

const char *pam_strerror(pam_handle_t *pamh, int errnum)
{
#ifdef UGLY_HACK_FOR_PRIOR_BEHAVIOR_SUPPORT  /* will be removed from v 1.0 */

    int possible_error;

    possible_error = (int) pamh;
    if (!(possible_error >= 0 && possible_error <= PAM_BAD_ITEM)) {
	possible_error = errnum;
    }

/* mask standard behavior to use possible_error variable. */
#define errnum possible_error

#endif /* UGLY_HACK_FOR_PRIOR_BEHAVIOR_SUPPORT */

    switch (errnum) {
    case PAM_SUCCESS:
	return "Success";
    case PAM_ABORT:
	return "Critical error - immediate abort";
    case PAM_OPEN_ERR:
	return "dlopen() failure";
    case PAM_SYMBOL_ERR:
	return "Symbol not found";
    case PAM_SERVICE_ERR:
	return "Error in service module";
    case PAM_SYSTEM_ERR:
	return "System error";
    case PAM_BUF_ERR:
	return "Memory buffer error";
    case PAM_PERM_DENIED:
	return "Permission denied";
    case PAM_AUTH_ERR:
	return "Authentication failure";
    case PAM_CRED_INSUFFICIENT:
	return "Insufficient credentials to access authentication data";
    case PAM_AUTHINFO_UNAVAIL:
	return "Authentication service cannot retrieve authentication info.";
    case PAM_USER_UNKNOWN:
	return "User not known to the underlying authentication module";
    case PAM_MAXTRIES:
	return "Have exhasted maximum number of retries for service.";
    case PAM_NEW_AUTHTOK_REQD:
	return "Authentication token is no longer valid; new one required.";
    case PAM_ACCT_EXPIRED:
	return "User account has expired";
    case PAM_SESSION_ERR:
	return "Cannot make/remove an entry for the specified session";
    case PAM_CRED_UNAVAIL:
	return "Authentication service cannot retrieve user credentials";
    case PAM_CRED_EXPIRED:
	return "User credentials expired";
    case PAM_CRED_ERR:
	return "Failure setting user credentials";
    case PAM_NO_MODULE_DATA:
	return "No module specific data is present";
    case PAM_BAD_ITEM:
	return "Bad item passed to pam_*_item()";
    case PAM_CONV_ERR:
	return "Conversation error";
    case PAM_AUTHTOK_ERR:
	return "Authentication token manipulation error";
    case PAM_AUTHTOK_RECOVER_ERR:
	return "Authentication information cannot be recovered";
    case PAM_AUTHTOK_LOCK_BUSY:
	return "Authentication token lock busy";
    case PAM_AUTHTOK_DISABLE_AGING:
	return "Authentication token aging disabled";
    case PAM_TRY_AGAIN:
	return "Failed preliminary check by password service";
    case PAM_IGNORE:
	return "Please ignore underlying account module";
    case PAM_MODULE_UNKNOWN:
	return "Module is unknown";
    case PAM_AUTHTOK_EXPIRED:
	return "Authentication token expired";
    case PAM_CONV_AGAIN:
	return "Conversation is waiting for event";
    case PAM_INCOMPLETE:
	return "Application needs to call libpam again";
    }

    return "Unknown Linux-PAM error (need to upgrde libpam?)";
}
